<div class="alert alert-block alert-info">
<h2> Комментарии </h2>
</div>

Привет, Михаил!

Я проверила твою работу и оставила в ней комментарии. Комментарии оформлены в таких ячейках и вставлены сразу после ячейки или группы ячеек, к которым относятся. Цвета ячеек означают:

<div class="alert alert-block alert-danger">
<p> Красные ячейки говорят о том, что нужно что-то исправить или пояснить. Для того, чтобы проект был принят, необходимо отреагировать на комментарии в этих ячейках. </p>
</div>
<div class="alert alert-block alert-info">
<p> Синие ячейки говорят о том, что в этой части работы все хорошо. Могут содержать рекомендаци, предложения, корректировки.</p>
</div>

Надеюсь, тебе все будет понятно :)

Хорошая работа 🤓. Ты делаешь правильные выводы и понимаешь зачем нужен каждый шаг 👍.

Отправляю проект на доработку, так как есть несколько ошибок и пара мест, требующие уточнения.

Читай комментарии, поправляй работу и возвращай на проверку.

***Буду ждать финальную версию работы!***

------------

<div class="alert alert-block alert-info">
<h2> Комментарии 1 </h2>
</div>

Привет!

Проект зачтен :) Читай комментарии внутри работы в таких ячейках.

***Успехов!!!***

------------

### Описание проекта
Вы — аналитик крупного интернет-магазина. Вместе с отделом маркетинга вы подготовили список гипотез для увеличения выручки.
Приоритизируйте гипотезы, запустите A/B-тест и проанализируйте результаты.

In [1]:
import pandas as pd
import numpy as np
import scipy.stats as stats
import seaborn as sns
import matplotlib.pyplot as plt
from collections import defaultdict
import datetime as dt
from plotly.offline import download_plotlyjs, init_notebook_mode, plot, iplot
import plotly.graph_objs as go

init_notebook_mode(connected=True)
pd.set_option('display.max_colwidth', -1)

### Задача 1

In [2]:
hypothesis = pd.read_csv('/datasets/hypothesis.csv')
hypothesis.columns = ['hypothesis', 'reach', 'impact', 'confidence', 'efforts']
hypothesis

hypothesis  \
0  Добавить два новых канала привлечения трафика, что позволит привлекать на 30% больше пользователей             
1  Запустить собственную службу доставки, что сократит срок доставки заказов                                      
2  Добавить блоки рекомендаций товаров на сайт интернет магазина, чтобы повысить конверсию и средний чек заказа   
3  Изменить структура категорий, что увеличит конверсию, т.к. пользователи быстрее найдут нужный товар            
4  Изменить цвет фона главной страницы, чтобы увеличить вовлеченность пользователей                               
5  Добавить страницу отзывов клиентов о магазине, что позволит увеличить количество заказов                       
6  Показать на главной странице баннеры с актуальными акциями и распродажами, чтобы увеличить конверсию           
7  Добавить форму подписки на все основные страницы, чтобы собрать базу клиентов для email-рассылок               
8  Запустить акцию, дающую скидку на товар в день рождения                                                        

   reach  impact  confidence  efforts  
0  3      10      8           6        
1  2      5       4           10       
2  8      3       7           3        
3  8      3       3           8        
4  3      1       1           1        
5  3      2       2           3        
6  5      3       8           3        
7  10     7       8           5        
8  1      9       9           5

##### Примените фреймворк ICE для приоритизации гипотез. Отсортируйте их по убыванию приоритета.

In [3]:
hypothesis['ICE'] = hypothesis['impact'] * hypothesis['confidence'] / hypothesis['efforts']
hypothesis_ICE = hypothesis[['hypothesis', 'ICE']].sort_values(by='ICE', ascending=False).round(2)
hypothesis_ICE

hypothesis  \
8  Запустить акцию, дающую скидку на товар в день рождения                                                        
0  Добавить два новых канала привлечения трафика, что позволит привлекать на 30% больше пользователей             
7  Добавить форму подписки на все основные страницы, чтобы собрать базу клиентов для email-рассылок               
6  Показать на главной странице баннеры с актуальными акциями и распродажами, чтобы увеличить конверсию           
2  Добавить блоки рекомендаций товаров на сайт интернет магазина, чтобы повысить конверсию и средний чек заказа   
1  Запустить собственную службу доставки, что сократит срок доставки заказов                                      
5  Добавить страницу отзывов клиентов о магазине, что позволит увеличить количество заказов                       
3  Изменить структура категорий, что увеличит конверсию, т.к. пользователи быстрее найдут нужный товар            
4  Изменить цвет фона главной страницы, чтобы увеличить вовлеченность пользователей                               

     ICE  
8  16.20  
0  13.33  
7  11.20  
6  8.00   
2  7.00   
1  2.00   
5  1.33   
3  1.12   
4  1.00

##### Примените фреймворк RICE для приоритизации гипотез. Отсортируйте их по убыванию приоритета.

In [4]:
hypothesis['RICE'] = hypothesis['reach'] * hypothesis['impact'] * hypothesis['confidence'] / hypothesis['efforts']
hypothesis_RICE = hypothesis[['hypothesis', 'RICE']].sort_values(by='RICE', ascending=False)
hypothesis_RICE

hypothesis  \
7  Добавить форму подписки на все основные страницы, чтобы собрать базу клиентов для email-рассылок               
2  Добавить блоки рекомендаций товаров на сайт интернет магазина, чтобы повысить конверсию и средний чек заказа   
0  Добавить два новых канала привлечения трафика, что позволит привлекать на 30% больше пользователей             
6  Показать на главной странице баннеры с актуальными акциями и распродажами, чтобы увеличить конверсию           
8  Запустить акцию, дающую скидку на товар в день рождения                                                        
3  Изменить структура категорий, что увеличит конверсию, т.к. пользователи быстрее найдут нужный товар            
1  Запустить собственную службу доставки, что сократит срок доставки заказов                                      
5  Добавить страницу отзывов клиентов о магазине, что позволит увеличить количество заказов                       
4  Изменить цвет фона главной страницы, чтобы увеличить вовлеченность пользователей                               

    RICE  
7  112.0  
2  56.0   
0  40.0   
6  40.0   
8  16.2   
3  9.0    
1  4.0    
5  4.0    
4  3.0

##### Укажите, как изменилась приоритизация гипотез при применении RICE вместо ICE. Объясните, почему так произошло.

In [5]:
priority_of_hypothesis = hypothesis_ICE.merge(hypothesis_RICE, on='hypothesis')
priority_of_hypothesis

hypothesis  \
0  Запустить акцию, дающую скидку на товар в день рождения                                                        
1  Добавить два новых канала привлечения трафика, что позволит привлекать на 30% больше пользователей             
2  Добавить форму подписки на все основные страницы, чтобы собрать базу клиентов для email-рассылок               
3  Показать на главной странице баннеры с актуальными акциями и распродажами, чтобы увеличить конверсию           
4  Добавить блоки рекомендаций товаров на сайт интернет магазина, чтобы повысить конверсию и средний чек заказа   
5  Запустить собственную службу доставки, что сократит срок доставки заказов                                      
6  Добавить страницу отзывов клиентов о магазине, что позволит увеличить количество заказов                       
7  Изменить структура категорий, что увеличит конверсию, т.к. пользователи быстрее найдут нужный товар            
8  Изменить цвет фона главной страницы, чтобы увеличить вовлеченность пользователей                               

     ICE   RICE  
0  16.20  16.2   
1  13.33  40.0   
2  11.20  112.0  
3  8.00   40.0   
4  7.00   56.0   
5  2.00   4.0    
6  1.33   4.0    
7  1.12   9.0    
8  1.00   3.0

Параметр охвата (Reach), использующийся в методе RICE изменил всю приоритетность, полученную по методу ICE, только 6 и 4 гипотезы остались на прежних местах.

<div class="alert alert-block alert-info">
<h2> Комментарии </h2>
</div>

Отлично :)

------------

### Задача 2

#### Предобработка данных

In [6]:
orders = pd.read_csv('/datasets/orders.csv')
visitors = pd.read_csv('/datasets/visitors.csv')

Данные оказались чистыми, проверим нет ли пользователей, которые оказались в обеих группах одновременно

In [7]:
group_sizes = orders.groupby(['visitorId', 'group']).size().reset_index()
group_sizes.columns = ['visitorId', 'group', 'count']
group_sizes = group_sizes.query('count > 1').sort_values(by='visitorId')
group_sizes

visitorId group  count
45    199603092   A     4    
58    237748145   A     4    
63    249864742   B     3    
94    366673373   A     2    
104   393266494   B     2    
...         ...  ..    ..    
1045  4109358064  A     2    
1050  4134830387  B     2    
1065  4186807279  B     2    
1079  4256040402  A     7    
1080  4256040402  B     4    

[64 rows x 3 columns]

In [8]:
potentially_broken_users_list = list(zip(group_sizes['visitorId'], group_sizes['group']))
potentially_broken_users_dict = defaultdict(list)
for user, group in potentially_broken_users_list:
    potentially_broken_users_dict[user].append(group)

inappropriate_users = [user for user in potentially_broken_users_dict if len(potentially_broken_users_dict[user]) > 1]
inappropriate_orders = orders.query('visitorId in @inappropriate_users')
inappropriate_orders_grouped = (inappropriate_orders
                        .groupby(['visitorId', 'group', 'date'])
                        .agg({'transactionId': 'nunique', 'revenue': 'sum'})
                   )
inappropriate_orders_grouped

transactionId  revenue
visitorId  group date                              
1230306981 A     2019-08-09  3              18440  
           B     2019-08-09  2              12840  
2038680547 A     2019-08-04  2              4070   
                 2019-08-17  1              3470   
                 2019-08-22  1              760    
                 2019-08-25  1              270    
           B     2019-08-04  1              2800   
                 2019-08-15  1              1680   
                 2019-08-29  1              2420   
2378935119 A     2019-08-10  1              58220  
                 2019-08-15  1              1650   
                 2019-08-17  1              310    
                 2019-08-22  1              3390   
           B     2019-08-09  1              30649  
                 2019-08-10  1              2640   
                 2019-08-25  1              21220  
                 2019-08-28  2              24860  
2458001652 A     2019-08-04  1              340    
                 2019-08-06  1              1620   
                 2019-08-07  1              17379  
                 2019-08-11  1              8280   
                 2019-08-22  1              5190   
           B     2019-08-05  1              2420   
                 2019-08-06  2              14030  
                 2019-08-08  1              5227   
                 2019-08-09  1              5227   
                 2019-08-25  1              2385   
4256040402 A     2019-08-01  1              5800   
                 2019-08-12  1              50     
                 2019-08-17  1              1550   
                 2019-08-20  1              17650  
                 2019-08-23  1              16950  
                 2019-08-27  1              1630   
                 2019-08-31  1              5170   
           B     2019-08-08  1              58220  
                 2019-08-16  1              56650  
                 2019-08-18  1              10970  
                 2019-08-31  1              1850

Пять пользователей оказались в двух группах одновременно, они совершили достаточно много транзакций и принесли значимую суму выручки, поэтому нужно очистить наши таблицы от этих пользователей, для корретного А/В теста

In [9]:
orders_filt = orders.query('visitorId not in @inappropriate_orders')

Теперь отфильтруем таблицу visitors

In [10]:
inappropriate_orders_grouped = inappropriate_orders_grouped.reset_index()
inappropriate_orders_grouped['count'] = -1
inappropriate_orders_grouped = inappropriate_orders_grouped[['group', 'date', 'count']]
visitors_filt = visitors.merge(inappropriate_orders_grouped, on=['date', 'group'], how='left').fillna(0)
visitors_filt['visitors_'] = visitors_filt['visitors'] + visitors_filt['count']
visitors_filt = visitors_filt.drop(['visitors', 'count'], axis=1)
visitors_filt.columns = ['date', 'group', 'visitors']
visitors_filt['visitors'] = visitors_filt['visitors'].astype(int)

<div class="alert alert-block alert-info">
<h2> Комментарии </h2>
</div>

Хорошо :)

Так как в рамках этого A/B теста не пользователей, а сесии разделяли на группы, то проблемы в том, что пользователь попал в обе группы, нет. Просто так вышло, что он совершил несколько сессий, которые определились в разные группы.

Пожалуста, не опускай шаги :) Хотелось бы увидеть весь анализ, который ты провел, чтобы понять не упустил ли ты что-то.

------------

##### 1) Постройте график кумулятивной выручки по группам. Сделайте выводы и предположения.

In [11]:
cum_revenue = orders_filt.groupby(['group', 'date'])['revenue'].sum().groupby(level=[0]).cumsum().to_frame()
cum_revenue = cum_revenue.reset_index()
cum_revenue['date'] = cum_revenue['date'].map(lambda x: dt.datetime.strptime(x, '%Y-%m-%d'))
cum_revenue_A = cum_revenue[cum_revenue['group'] == 'A'][['date', 'revenue']]
cum_revenue_B = cum_revenue[cum_revenue['group'] == 'B'][['date', 'revenue']]
cum_revenue

group       date  revenue
0   A    2019-08-01  148579 
1   A    2019-08-02  242401 
2   A    2019-08-03  354874 
3   A    2019-08-04  425699 
4   A    2019-08-05  549917 
.. ..           ...     ... 
57  B    2019-08-27  5232260
58  B    2019-08-28  5328268
59  B    2019-08-29  5559398
60  B    2019-08-30  5774631
61  B    2019-08-31  5907760

[62 rows x 3 columns]

In [12]:
trace_A = go.Scatter(
    x = cum_revenue_A['date'],
    y = cum_revenue_A['revenue'],
    mode = 'lines+markers',
    name = 'A'
)
trace_B = go.Scatter(
    x = cum_revenue_B['date'],
    y = cum_revenue_B['revenue'],
    mode = 'lines+markers',
    name = 'B'
)

layout = go.Layout(
    title='Кумулятивная выручка по группам',
)   

fig = go.Figure(data = [trace_A, trace_B], layout = layout)
iplot(fig)

##### 2) Постройте график кумулятивного среднего чека по группам. Сделайте выводы и предположения.

In [13]:
cum_mean_check = orders_filt.groupby(['group','date'])['revenue'].mean().groupby(level=[0]).cumsum().to_frame().reset_index().round(1)
cum_mean_check
cum_mean_check_A = cum_mean_check[cum_mean_check['group']=='A']
cum_mean_check_B = cum_mean_check[cum_mean_check['group']=='B']


In [14]:
trace_A = go.Scatter(
    x = cum_mean_check_A['date'],
    y = cum_mean_check_A['revenue'],
    mode = 'lines+markers',
    name = 'A'
)
trace_B = go.Scatter(
    x = cum_mean_check_B['date'],
    y = cum_mean_check_B['revenue'],
    mode = 'lines+markers',
    name = 'B'
)

layout = go.Layout(
    title='Кумулятивный средний чек по группам',
    xaxis_title='Дата',
    yaxis_title='Средний чек'
)   

fig = go.Figure(data = [trace_A, trace_B], layout = layout)
iplot(fig)

### Исправленный вариант графика кумулятивного среднего чека и графика относительного изменения, но вопрос остался, почему данные разнятся, не могу понять где именно расходятся коды, подскажи пожалуйста, почему мой вариант не подходит, совсем недоходит до меня))

In [15]:
datesGroups = orders[['date','group']].drop_duplicates()
ordersAggregated = datesGroups.apply(lambda x: orders
                                     [np.logical_and(orders['date'] <= x['date'],
                                                     orders['group'] == x['group'])].agg({'date' : 'max',
                                                                                          'group' : 'max', 'transactionId' : pd.Series.nunique,
                                                                                          'visitorId' : pd.Series.nunique, 'revenue' : 'sum'}), axis=1).sort_values(by=['date','group'])

visitorsAggregated = datesGroups.apply(lambda x: visitors
                                       [np.logical_and(visitors['date'] <= x['date'],
                                                       visitors['group'] == x['group'])].agg({'date' : 'max',
                                                                                              'group' : 'max',
                                                                                              'visitors' : 'sum'}), axis=1).sort_values(by=['date','group'])






In [16]:
cummulativeData = ordersAggregated.merge(visitorsAggregated, left_on=['date', 'group'], right_on=['date', 'group'])
cummulativeData.columns = ['date', 'group', 'orders', 'buyers', 'revenue', 'visitors']

In [17]:
cummulativeRevenueA = cummulativeData[cummulativeData['group']=='A'][['date','revenue', 'orders']]
cummulativeRevenueB = cummulativeData[cummulativeData['group']=='B'][['date','revenue', 'orders']]
trace_A = go.Scatter(
    x = cummulativeRevenueA['date'],
    y = cummulativeRevenueA['revenue']/cummulativeRevenueA['orders'],
    mode = 'lines+markers',
    name = 'A'
)
trace_B = go.Scatter(
    x = cummulativeRevenueB['date'],
    y = cummulativeRevenueB['revenue']/cummulativeRevenueB['orders'],
    mode = 'lines+markers',
    name = 'B'
)

layout = go.Layout(
    title='Кумулятивный средний чек по группам',
    xaxis_title='Дата',
    yaxis_title='Средний чек'
)   

fig = go.Figure(data = [trace_A, trace_B], layout = layout)
iplot(fig)

In [18]:
mean_check_A = cummulativeData.query('group == "A"')['revenue'].tolist()
mean_check_B = cummulativeData.query('group == "B"')['revenue'].tolist()
mean_check_rel = [y/x - 1 for x,y in zip(mean_check_A, mean_check_B)]

cum_mean_check_rel = pd.DataFrame({'date': cum_mean_check['date'].unique(),
                                   'mean_check_rel': mean_check_rel})
layout = go.Layout(title='График относительного изменения кумулятивного среднего чека',
                   xaxis_title='Дата',
                   yaxis_title='Средний чек')
fig = go.Figure(layout=layout)
fig.add_trace(go.Scatter(x=cum_mean_check_rel['date'],
                                      y=cum_mean_check_rel['mean_check_rel'],
                                      mode='lines+markers', line = dict(color='black')
                                     ))
fig.add_trace( go.Scatter(x = [min(cum_mean_check_rel['date']), max(cum_mean_check_rel['date'])], y=[0, 0], mode="lines",
                        line = dict(color='royalblue', width=4, dash='dash'))) ## <------------------------------

fig.show()

##### 3) Постройте график относительного изменения кумулятивного среднего чека группы B к группе A. Сделайте выводы и предположения.

In [19]:
mean_check_A = cum_mean_check.query('group == "A"')['revenue'].tolist()
mean_check_B = cum_mean_check.query('group == "B"')['revenue'].tolist()
mean_check_rel = [y/x - 1 for x,y in zip(mean_check_A, mean_check_B)]

cum_mean_check_rel = pd.DataFrame({'date': cum_mean_check['date'].unique(),
                                   'mean_check_rel': mean_check_rel})

In [20]:
layout = go.Layout(title='График относительного изменения кумулятивного среднего чека',
                   xaxis_title='Дата',
                   yaxis_title='Средний чек')
fig = go.Figure(layout=layout)
fig.add_trace(go.Scatter(x=cum_mean_check_rel['date'],
                                      y=cum_mean_check_rel['mean_check_rel'],
                                      mode='lines+markers', line = dict(color='black')
                                     ))
#fig.add_trace(go.Scatter(x=1, y=1,line = dict(color='royalblue', width=4, dash='dash')))

## --------------------------> линия параллельно оси х
fig.add_trace( go.Scatter(x = [min(cum_mean_check_rel['date']), max(cum_mean_check_rel['date'])], y=[0, 0], mode="lines",
                        line = dict(color='royalblue', width=4, dash='dash'))) ## <------------------------------

fig.show()

###### Вывод по пунткам 1-3 
На большей части графика средний чек и выручка по группе B выше, чем по группе A. До 18 августа они были практически наравне, но потом произошел резкий скачок у группы B, именно засчет этого скачка средний чек лежит выше, потому что в дальнейшем рост графиков идет одинаковыми темпами 

<div class="alert alert-block alert-danger">
<h2> Комментарии </h2>
</div>

Хорошо :)

В расчете кумулятивного среднего чека есть ошибка 😓. Здесь средний чек должен считаться как отношение кумулятивной выручки к кумулятивному количеству пользователей.

Как думаешь, за счет чего мог произойти скачок 18 августа в группе В?

Прости, что не ответила сразу в чате по поводу нанесения на график линии 😓. На последнем графике относительных измененний добавила код для это линии (линия выведена на графике). Для того, чтобы построить линию нужно аналагичным способом, как ты добавлял сам график, добавить график линии параллельной оси х через Scatter и задать значения x и y. Так как эта линия параллельна оси х и проходит через весь график, то по оси х она должна изменяться от минимального до максимального значения даты, а по оси у изменяться не должна, то есть должна быть зафиксирована в одной точке (я зафиксировала на точке 0, то есть изменение у идет от 0 до 0). Если остались вопросы, пиши :)

Не забывай подписывать оси графиков :)

------------

<div class="alert alert-block alert-info">
<h2> Комментарии 1 </h2>
</div>

Во второй версии у тебя получился не кумулятивный средний чек, а кумулятивная выручка :(

Кумулятивный средний чек для каждого дня нужно рассчитывать как отношение суммарной выручки на этой дате к общему количеству покупателей к этой дате. Ты же его рассчитываешь как сумму всех средних чеков к этой дате.


------------

##### 4) Постройте график кумулятивной конверсии по группам. Сделайте выводы и предположения.

In [21]:
orders_conversion = orders_filt.groupby(['date', 'group'])['transactionId'].nunique().to_frame().reset_index()
df_conversion = orders_conversion.merge(visitors_filt, on=['date', 'group'])
df_conversion['conversion'] = (df_conversion['transactionId'] / df_conversion['visitors']).round(4)
df_conversion = df_conversion[['date', 'group', 'conversion']]
df_conversion = (df_conversion
                          .groupby(['group', 'date'])['conversion']
                          .mean()
                          .groupby(level=[0])
                          .cumsum()
                          .round(2)
                          .to_frame()
                 )
df_conversion = df_conversion.reset_index()
df_conversion['date'] = df_conversion['date'].map(lambda x: dt.datetime.strptime(x, '%Y-%m-%d'))

In [22]:
df_conversion_A = df_conversion[df_conversion['group'] == 'A']
df_conversion_B = df_conversion[df_conversion['group'] == 'B']
layout = go.Layout(title='График кумулятивной конверсии по группам',
                  xaxis_title = 'Дата',
                  yaxis_title = 'Конверсия')

fig = go.Figure(layout=layout)
fig.add_trace(go.Scatter(x = df_conversion_A['date'],
                         y = df_conversion_A['conversion'], mode = 'lines+markers', name='A'))
fig.add_trace(go.Scatter(x = df_conversion_B['date'],
                         y = df_conversion_B['conversion'], mode = 'lines+markers', name='B'))
fig.show()

###### 5) Постройте график относительного изменения кумулятивной конверсии группы B к группе A. Сделайте выводы и предположения.

In [23]:
rel_conv = df_conversion
rel_conv_A = rel_conv.query('group == "A"')['conversion'].tolist()
rel_conv_B = rel_conv.query('group == "B"')['conversion'].tolist()
mean_conv_rel = [y/x - 1 for x,y in zip(rel_conv_A, rel_conv_B)]

cum_mean_conv_rel = pd.DataFrame({'date': rel_conv['date'].unique(),
                                  'conversion': mean_conv_rel})
cum_mean_conv_rel
layout = go.Layout(title = 'График относительного изменения кумулятивной конверсии группы B к группе A',
                  xaxis_title = 'Дата',
                  yaxis_title = 'Конверсия')
fig = go.Figure(layout=layout)
fig.add_trace(go.Scatter(x = cum_mean_conv_rel['date'],
                         y = cum_mean_conv_rel['conversion'],
                         mode='lines+markers', line = dict(color='black')))
fig.show()

###### Вывод по пунткам 4-5
По графику накопительная конверсия по группе B практически на всем протяжении лежит значительно выше накопительной конверсии группы A. Начиная с 4 августа вверх берет группа В и стабилизируется с 23 августа.

<div class="alert alert-block alert-info">
<h2> Комментарии </h2>
</div>

Отлично :)

------------

##### 6) Постройте точечный график количества заказов по пользователям. Сделайте выводы и предположения.

In [24]:
orders_by_users = orders_filt.groupby('visitorId', as_index=False).agg({'transactionId':'nunique'})
orders_by_users['count'] = range(1, len(orders_by_users)+1)

In [25]:
layout = go.Layout(title = 'Точечный график количества заказов по пользователям',
                  xaxis_title = 'Пользователи',
                  yaxis_title = 'Кол-во заказов')
fig = go.Figure(layout=layout)
fig.add_trace(go.Scatter(x=orders_by_users['count'], y=orders_by_users['transactionId'], mode='markers'))
fig.show()

В основном пользватели делают 1 заказ, меньшее кол-во поьзователей делает 2 заказа, но их все же много. Есть выбросы где пользователи делают 8,9, 11 заказов.

##### 7) Посчитайте 95-й и 99-й перцентили количества заказов на пользователя. Выберите границу для определения аномальных пользователей.

In [26]:
percentile_orders = np.percentile(orders_by_users['transactionId'], [95, 99])
percentile_orders

array([2., 4.])

99-му персентилю соотвествует значение 4, то есть больше 4 заказов сделали меньше 1% пользователей, 95-му персентилю соотвестует число 2, то есть больше 2 заказов совершили меньше 5% пользователей.

<div class="alert alert-block alert-danger">
<h2> Комментарии </h2>
</div>

Хорошо :)

Какова граница для определения аномальных пользователей?

------------

##### 8) Постройте точечный график стоимостей заказов. Сделайте выводы и предположения.

In [27]:
revenue_by_users = orders_filt
revenue_by_users['count'] = range(1, len(revenue_by_users)+1)

In [28]:
layout = go.Layout(title = 'Точечный график количества заказов по пользователям',
                  xaxis_title = 'Пользователи',
                  yaxis_title = 'Кол-во заказов')
fig = go.Figure(layout=layout)
fig.add_trace(go.Scatter(x=revenue_by_users['count'], y=revenue_by_users['revenue'], mode='markers'))
fig.show()

Наблюдатся огромный выброс на одного пользователя со стоимость заказа в 1.29 млн, построим нижем график с ограничем по У-оси.

In [29]:
layout = go.Layout(title = 'Точечный график количества заказов по пользователям',
                  xaxis_title = 'Пользователи',
                  yaxis_title = 'Кол-во заказов', yaxis=dict(range=[0,205000]))
fig = go.Figure(layout=layout)
fig.add_trace(go.Scatter(x=revenue_by_users['count'], y=revenue_by_users['revenue'], mode='markers'))

fig.show()

Видим, что в оссновном люди заказвают примерно до 10 тысяч рублей. Так же на графике красная точка ялвяется 95 персентилем в размере 28 тысяч

##### 9) Посчитайте 95-й и 99-й перцентили стоимости заказов. Выберите границу для определения аномальных заказов.


In [30]:
percentile_revenue = np.percentile(revenue_by_users['revenue'], [95,99])
percentile_revenue

array([28000. , 58233.2])

99-му персентилю соотвествует значение 58233, то есть стоимость заказа больше 58233 у 1% пользователей, 95-му персентилю соотвестует число 28000, то есть сумма заказа больше 28 тысяч у 5% пользователей

<div class="alert alert-block alert-danger">
<h2> Комментарии </h2>
</div>

Хорошо :)

Какова граница для определения аномальных пользователей?

------------

#### Сперва сформируем очищенный датафрем, где:
Возьмем 95-й персентиль из каждого значения.

1) Кол-во заказов на одного пользователя не более 2.

2) Сумма одного заказа не более 28000.

<div class="alert alert-block alert-danger">
<h2> Комментарии </h2>
</div>

Почему выбираешь именно такие значения для определения аномалий? Каким персентилям они соответсвуют?

------------

<div class="alert alert-block alert-info">
<h2> Комментарии 1 </h2>
</div>

Отлично :)

------------

In [31]:
clean_count_orders = orders_filt.groupby('visitorId')['transactionId'].count()
clean_count_orders = clean_count_orders[clean_count_orders > 2].index.tolist()

In [32]:
clean_sum_orders = orders_filt.groupby('visitorId')['revenue'].max()
clean_sum_orders = clean_sum_orders[clean_sum_orders > 28000].index.tolist()

In [33]:
orders_common = list(set(clean_count_orders + clean_sum_orders))

In [34]:
orders_final = orders_filt.query('visitorId not in @ orders_common')

<div class="alert alert-block alert-info">
<h2> Комментарии </h2>
</div>

Хорошо :)

------------

##### 10) Посчитайте статистическую значимость различий в конверсии между группами по «сырым» данным. Сделайте выводы и предположения.
##### 12) Посчитайте статистическую значимость различий в конверсии между группами по «очищенным» данным. Сделайте выводы и предположения.

Для расчета р-value, я решил использовать напараметрический тест Уилкоксона-Манна-Уитни, так как у нас имеются выбросы, а сам тест работает с рангами, не учитывая сами значения. Так же сформирую нулевую и альтернативную гипотезы. Нулевая гипотеза говорит о том, что статистическая значимость не значительна  и сделать вывод о различии нельзя, альтернативная гипотеза говорит о том, что статистическая значимость значитаельная и между выборками имеется статистическая значимость. Порогом статистической значимости установим alpha=.05

In [35]:
#p-value  по сырым данным
orders_conversion = orders_filt.groupby(['date', 'group'])['visitorId'].nunique().to_frame().reset_index()

df_conv_comparison = orders_conversion.merge(visitors, on=['date', 'group'])
df_conv_comparison['conversion'] = (df_conv_comparison['visitorId'] / df_conv_comparison['visitors']).round(4)
df_conv_comparison = df_conv_comparison[['date', 'group', 'conversion']]

conv_A = df_conv_comparison.query('group == "A"')['conversion'].tolist()
conv_B = df_conv_comparison.query('group == "B"')['conversion'].tolist()
conv_A_mean = sum(conv_A) / len(conv_A)
conv_B_mean = sum(conv_B) / len(conv_B)

#p-value - по очищенным данным
orders_conversion_clean = orders_final.groupby(['date', 'group'])['visitorId'].nunique().to_frame().reset_index()

df_conv_comparison_clean = orders_conversion_clean.merge(visitors, on=['date', 'group'])
df_conv_comparison_clean['conversion'] = (df_conv_comparison_clean['visitorId']/df_conv_comparison_clean['visitors'])
df_conv_comparison_clean['conversion'] = df_conv_comparison_clean['conversion'].round(4)
df_conv_comparison_clean = df_conv_comparison_clean[['date', 'group', 'conversion']]

conv_clean_A = df_conv_comparison_clean.query('group == "A"')['conversion'].tolist()
conv_clean_B = df_conv_comparison_clean.query('group == "B"')['conversion'].tolist()
conv_A_mean_cleared = sum(conv_clean_A)/len(conv_clean_A)
conv_B_mean_cleared = sum(conv_clean_B)/len(conv_clean_B)
alpha = .05
p_val_conv = stats.mannwhitneyu(conv_A, conv_B)[1].round(5)
p_val_conv_clean = stats.mannwhitneyu(conv_clean_A, conv_clean_B)[1].round(5)
print('P-value по конверсии до очистки данных: {}'.format(p_val_conv))
if (p_val_conv < alpha):
    print("Отвергаем нулевую гипотезу")
else:
    print("Не получилось отвергнуть нулевую гипотезу")
print('Отношения средних B/A по конверсии (до очистки данных): {:.2f}'.format(conv_B_mean/conv_A_mean))
print('P-value по конверсии после очистки данных: {}'.format(p_val_conv_clean))
if (p_val_conv_clean < alpha):
    print("Отвергаем нулевую гипотезу")
else:
    print("Не получилось отвергнуть нулевую гипотезу")
print('Отношения средних B/A по конверсии (после очистки данных): {:.2f}'.format
                                                                          (conv_B_mean_cleared/conv_A_mean_cleared))

P-value по конверсии до очистки данных: 0.03745
Отвергаем нулевую гипотезу
Отношения средних B/A по конверсии (до очистки данных): 1.16
P-value по конверсии после очистки данных: 0.0148
Отвергаем нулевую гипотезу
Отношения средних B/A по конверсии (после очистки данных): 1.19


P-value меньше 0.05 в обеих случаях. Значит, мы отвергаем нулевую гипотезу и подтверждаем, что в среднем чеке между группами есть статистически значимая разница. Что касается отношений средних чеков, то группа В стабильно выше группы А, до очистки данных выше на 16%, а после очистки данных, на 19%.

<div class="alert alert-block alert-danger">
<h2> Комментарии </h2>
</div>

Хорошо :)

Перед проведением тестов необходимо сформулировать нулевую гипотезу :) Почему используешь именно этот тест для проверки гипотезы? Какой выбираешь уровень значимости?

Поясни, пожалуйста, что значит фраза "...а после очистки данных разница возросла почти в 2 раза" 🤔. p-value не говорит о разнице между группами. Этот показатель отражает ошибку первого рода, то есть вероятность отвергнуть нулевую гипотезу, когда она верна.

Когда пишешь вывод по результатам теста, необходимо указывать на каком уровне значимости нулевая гипотеза отвергается или не отвергается, т.е., например, "на 5% уровне значимости нулевая гипотеза отвергается".   

------------

<div class="alert alert-block alert-info">
<h2> Комментарии 1 </h2>
</div>

Отлично :)

Нулевая гипотеза формулируется немного не так. Например, H0: конверсии в группах А и В по очищенным данным не отличаются/ равны. Здесь должны быть 2 нулевые гипотезы: одна для "очищенных данных", одна для "сырых данных".

------------

##### 11) Посчитайте статистическую значимость различий в среднем чеке заказа между группами по «сырым» данным. Сделайте выводы и предположения.

##### 13) Посчитайте статистическую значимость различий в среднем чеке заказа между группами по «очищенным» данным. Сделайте выводы и предположения.

In [36]:
mean_check_A = orders_filt[orders_filt['group'] == 'A']['revenue']
mean_check_B = orders_filt[orders_filt['group'] == 'B']['revenue']
mean_check_A_final = sum(mean_check_A) / len(mean_check_A)
mean_check_B_final = sum(mean_check_B) / len(mean_check_B)

mean_check_clean_A = orders_final[orders_final['group'] == 'A']['revenue']
mean_check_clean_B = orders_final[orders_final['group'] == 'B']['revenue']
mean_check_clean_A_final = sum(mean_check_clean_A) / len(mean_check_clean_A)
mean_check_clean_B_final = sum(mean_check_clean_B) / len(mean_check_clean_B)

p_val_mean_check_clean = stats.mannwhitneyu(mean_check_clean_A, mean_check_clean_B)[1].round(2)
p_val_mean_check = stats.mannwhitneyu(mean_check_A, mean_check_B)[1].round(2)
alpha = .05
print('P-value по конверсии до очистки данных: {}'.format(p_val_mean_check))
if (p_val_mean_check < alpha):
    print("Отвергаем нулевую гипотезу")
else:
    print("Не получилось отвергнуть нулевую гипотезу")
print('Отношения средних B/A по конверсии (до очистки данных): {:.2f}'.format(mean_check_B_final/mean_check_A_final))
print('P-value по конверсии после очистки данных: {}'.format(p_val_mean_check_clean))
if (p_val_mean_check_clean < alpha):
    print("Отвергаем нулевую гипотезу")
else:
    print("Не получилось отвергнуть нулевую гипотезу")
print('Отношения средних B/A по конверсии (после очистки данных): {:.2f}'.format
                                                                          (mean_check_clean_B_final/mean_check_clean_A_final))



P-value по конверсии до очистки данных: 0.36
Не получилось отвергнуть нулевую гипотезу
Отношения средних B/A по конверсии (до очистки данных): 1.26
P-value по конверсии после очистки данных: 0.38
Не получилось отвергнуть нулевую гипотезу
Отношения средних B/A по конверсии (после очистки данных): 0.98


В обеих случаях, такую разницу можно получить с вероятностью в 36%(до очистки данных) и в 38%(после очистки даных) поэтому мы принимаем нулевую гипотезу и говорим о том, что статистической значимости между выборками нет. Что касается отношений средних чеков, после очистки данных. Что касается среднего чека, то после очистки данных преимущество группы В упало на 28% и на 2% средний чек выше у группы А.

Напиши пожалуйста один комментарий по выбору, как правильно описать данную проверку, мне сложно вникнуть в связь между р-value по конверсии и например отношением средних чеков. У них есть какая то связь? И как правильнее доносить числа на словах,  в этом у меня тоже сложности((

<div class="alert alert-block alert-danger">
<h2> Комментарии </h2>
</div>

Хорошо :)

Комментарии аналогичны тем, что относятся к предыдущей группе тестов. Сформулируй, пожалуйста, нулевую гипотезу, обоснуй выбор теста и укажи уровень значимости :)

------------

<div class="alert alert-block alert-info">
<h2> Комментарии 1 </h2>
</div>

Отлично :)

Здесь тоже нужны нулевые гипотезы :)

Оба варианта, которыми ты описал результаты теста (для тестов по конверсии и для тестов по средним чекам), правильные, в них нет ошибок 🤓 Более распространенным вариантом описания результатов является тот, которым ты описываешь результаты тестов по конверсии.

Связи у p-value по конверсии и отношения средних чеков нет :)

------------

##### Результаты А/В теста:
- В конверссии между группами А и В есть статистическая значимость различий как по сырым данным, так и по очищенным.
- По средним чекам между группами А и В нет статистической значимости различий как по сырым данным, так и по очищенным.

## Общий вывод

Тест можно остановить и считать успешно пройденным. Так как после анализа мы видим, что группа В является более успешной, конверсия из пользователя в покупателя у группы В выше чем у группы А, хотя конверсия по среднему чеку не имеет значимых различий между группами. 

<div class="alert alert-block alert-info">
<h2> Комментарии </h2>
</div>

Отлично :)

------------